core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -4690,6 +4723,62 @@ object Stream extends StreamLowPriority {
       new Pull(Algebra.stepLeg(self))
   }
 
+  /**
+    *  Implementation for parZip. `AnyVal` classes do not allow inner
+    *  classes, so the presence of the `State` trait forces this
+    *  method to be outside of the `Stream` class.
+    */
+  private[fs2] def parZip[F[_]: Concurrent, L, R](
+      left: Stream[F, L],
+      right: Stream[F, R]
+  ): Stream[F, (L, R)] = {
+    sealed trait State
+    case object Racing extends State
+    case class LeftFirst(leftValue: L, waitOnRight: Deferred[F, Unit]) extends State
+    case class RightFirst(rightValue: R, waitOnLeft: Deferred[F, Unit]) extends State
+
+    def emit(l: L, r: R) = Pull.output1(l -> r)
+
+    Stream.eval(Ref[F].of(Racing: State)).flatMap { state =>
+      def lhs(stream: Stream[F, L]): Pull[F, (L, R), Unit] =
+        stream.pull.uncons1.flatMap {
+          case None => Pull.done
+          case Some((leftValue, stream)) =>
+            Pull.eval {
+              Deferred[F, Unit].flatMap { awaitRight =>
+                state.modify {
+                  case Racing =>
+                    LeftFirst(leftValue, awaitRight) -> Pull.eval(awaitRight.get)
+                  case RightFirst(rightValue, awaitLeft) =>
+                    Racing -> { emit(leftValue, rightValue) >> Pull.eval(awaitLeft.complete(())) }
+                  case LeftFirst(_, _) => sys.error("fs2 parZip protocol broken by lhs. File a bug")
```

Comment:
Would it be worth spitting out the left/right values in the error messages in these cases?

Reply:
I don't know. Generally how would so in an error case, but this would not be an error, it would be a bug, since I'm pretty confident it cannot possibly happen, it only needs to be there for pattern matching completeness